### PR TITLE
Tickets/dm 33163

### PR DESCRIPTION
**Pull request by teweitsai**
Initial migration of the common codes from `ts_rotator_controller`.

### CODE REVIEW

File: doc/commandStatus.md
Comment (r-owen) [inline doc/commandStatus.md]: Do we still need this sync pattern? Switching to common code seems like a nice time to take it out, if not.

File: doc/commandStatus.md
Comment (r-owen) [inline doc/commandStatus.md]: This seems to imply that only DDS gets command ack, but I hope both do. Please consider a slightly different wording, such as the following (assuming I'm right about the way the EUI takes control):

"DDS issues a command when the EUI has control. (The EUI will not see this because the EUI takes control from DDS whenever it issues a command.)"

File: doc/commandStatus.md
Comment (r-owen) [inline doc/commandStatus.md]: Surely this is outdated? You are holding off issuing the OK until the Simulink state changes, which is far more rosy than the picture you paint here. Please consider something like the following, if it's true (as I hope it is):

Most commands will get an OK when they have finished. The commands that take a long time to finish get an OK when they have successfully started. Those are:
* stop: OK when the enabled substate transitions to STOPPING or STOPPED
* moving point to point: OK when the enabled substate transitions to MOVING POINT TO POINT
* ...(fault? constant-velocity move for the rotator, any others?)

File: include/circular_buffer.h
Comment (r-owen) [inline include/circular_buffer.h]: It might be a bit clearer to say:
// Add data to the buffer. Delete old data to make room, if necessary.
// Return true if old data was deleted, else false.

File: include/circular_buffer.h
Comment (r-owen) [inline include/circular_buffer.h:43]: Please clarify that this removes the data from the queue and that it returns the oldest data. 
Also please clarify that it returns true if the buffer *was* already empty, so that no data was returned. Your text implies that it returns false if the buffer is now empty, even if the call returned the last remaining data.
Perhaps something like this:

// Pop and return the oldest data from the buffer.
// Return true if the buffer was empty (no data returned), false otherwise.

File: include/configPxi.h
Comment (r-owen) [inline include/configPxi.h]: Please delete the word "rotator". I suggest a global search for this.

File: include/interface/cmdServer.h
Comment (r-owen) [inline include/interface/cmdServer.h]: I gently suggest listing each parameter on its own line, because there are so many. That will make it easier to read the parameter documentation.

File: include/interface/cmdServer.h
Comment (r-owen) [inline include/interface/cmdServer.h]: Please consider returning a bool for these functions, just like your command buffer functions, to make the library more self-consistent.

An int is useful if you have to return multiple different error values, but at this point I suspect you will not.

File: include/interface/commandStructure.h
Comment (r-owen) [inline include/interface/commandStructure.h]: Is this still useful/needed? I'd gladly update my code to stop sending it if you can manage without it.

File: include/interface/commandStructure.h
Comment (r-owen) [inline include/interface/commandStructure.h]: Please delete this entry and the associated value. Any value other than the two below are cause for error.

File: include/interface/tlmServer.h
Comment (r-owen) [inline include/interface/tlmServer.h]: Please do a global search for "Many compilers assign the first element in an enumerated type" and delete it. It is meaningless when you explicitly assign values (as you always do), and thus confusing to readers.

In addition: if you don't use the "Unknown" enum name in your code then I strongly suggest deleting it.

You are being fully safe by avoiding 0 as a meaningful value.

File: include/interface/commandStructure.h
Comment (r-owen) [inline include/interface/commandStructure.h]: If you don't use this in your code then I suggest deleting it.

File: tests/testUtility.cpp
Comment (r-owen) [inline tests/testUtility.cpp:61]: This seems specific to the rotator. Can you please add some comments explaining how that works here. At least say that these mask values come from the MT Rotator. Or are just made up, if that's the case.

File: tests/testData/default.yaml
Comment (r-owen) [inline tests/testData/default.yaml]: Any chance of changing DDS to CSC now, in this common package? This low-level controller does not speak DDS at all! And we may switch away from DDS to some other messaging system, which will make this even more confusing in the future.

**Review by r-owen — CHANGES_REQUESTED** [review]
This looks great. I am very pleased that you are factoring out the common code.

My main requests are:
* Remove references to the rotator for things that apply to the hexapod as well
* Remove the enum comments about compilers assigning 0. Also I strongly suggest removing the Unknown entries unless you are using them. By assigning non-zero values explicitly you completely avoid the issue in question.

I have a few other suggestion I hope you will consider.

File: tests/testData/default.yaml
Comment (teweitsai) [reply tests/testData/default.yaml]: I removed these in the commit: 0f7015f. I would like to redesign the existed codes in DM-33252. Thanks!

File: tests/testUtility.cpp
Comment (teweitsai) [reply tests/testUtility.cpp:61]: This is for the rotator and hexapod. The hardware engineer did the same modification for the rotator, camera hexapod, and M2 hexapod. Thanks!

File: include/interface/commandStructure.h
Comment (teweitsai) [reply include/interface/commandStructure.h]: Removed it in commit: 05f3610. The other enum values had been changed as well. Thanks!

File: include/interface/tlmServer.h
Comment (teweitsai) [reply include/interface/tlmServer.h]: I can do this but I just feel if you do not like the suggestion from the author of "Code Complete 2", you could write the email to discuss with him. I just followed the textbook to do so. I felt such kind of argument had became the personal taste that is meaningless. Thanks!

File: include/interface/commandStructure.h
Comment (teweitsai) [reply include/interface/commandStructure.h]: I removed this enum in commit: 05f3610. Thanks!

File: include/configPxi.h
Comment (teweitsai) [reply include/configPxi.h]: Done in the commit : 0f7015fc. Thanks!

File: include/interface/cmdServer.h
Comment (teweitsai) [reply include/interface/cmdServer.h]: I removed these in the commit: 0f7015f. I would like to redesign the existed codes in DM-33252. Thanks!

File: include/interface/commandStructure.h
Comment (teweitsai) [reply include/interface/commandStructure.h]: Removed in the commit: 05f3610. Thanks!

File: include/interface/cmdServer.h
Comment (teweitsai) [reply include/interface/cmdServer.h]: I removed these in the commit: 0f7015f. I would like to redesign the existed codes in DM-33252. Thanks!

File: doc/commandStatus.md
Comment (teweitsai) [reply doc/commandStatus.md]: Use the `commander` instead of sync pattern. Update the document in commit: fd16057. Thanks!

File: doc/commandStatus.md
Comment (teweitsai) [reply doc/commandStatus.md]: Improved the document in commit: fd16057. Thanks!

File: doc/commandStatus.md
Comment (teweitsai) [reply doc/commandStatus.md]: Improved the document in commit: fd16057. Thanks!

File: include/circular_buffer.h
Comment (teweitsai) [reply include/circular_buffer.h]: Improved the document in commit: fd16057. Thanks!

File: include/circular_buffer.h
Comment (teweitsai) [reply include/circular_buffer.h:43]: Improved the document in commit: fd16057. Thanks!

File: include/interface/tlmServer.h
Comment (r-owen) [reply include/interface/tlmServer.h]: The comment can only be relevant if you don't assign some of the enum values yourself. But you always do. I find it very confusing to read comments about compiler behavior that cannot possibly be relevant to the code.

Please check that section of the book again. You will probably not be surprised to hear that I'm not a fan of that book (though I don't remember which edition I read). But even so, I'd be amazed if they suggest such a comment in the case that you assign all enum values yourself.

File: doc/commandStatus.md
Comment (r-owen) [inline doc/commandStatus.md]: The reality is that 0 often means "already done". So this needs some rewording.

Please consider making 0 *always* mean "finished". To do this, you would need to ensure that any code that estimates duration (for an OK ack for a command that is not done yet) always returns a positive value. It's might require a bit more code, but I think it would be worth it so clients can tell the difference between "done" and "started".

(If you dislike overloading `duration` in that way, another option is to add a boolean "done" flag to command status.)

File: include/interface/tlmServer.h
Comment (teweitsai) [reply include/interface/tlmServer.h]: I am pretty sure the example of book assigns all the values in enum.

File: doc/commandStatus.md
Comment (r-owen) [inline doc/commandStatus.md]: Please fill in this list yourself or delete this line.

It is probably safest to just provide a few obvious examples. But if the complete list is short and unlikely to change as the rotator and hexapods evolve, then it is more helpful to provide have the complete list here (or the rotator-specific list in the rotator docs and the hexapod-specific list in the hexapod docs).

`Point-to-point move` is a great example and it applies to both the rotator and the hexapod.

Would `stop` be another good example (and does it apply to hexapods, despite the fact that they stop very quickly)?

What about `fault`? Is it acked when done or when started?

Please consider deleting "(using the rotator as an example)" at line 37, because:
* If you list a few examples then most of them are likely to apply to both, and you can call out those that do not.
* If you list all commands then I think you have no choice but to call out which devices they apply to, when they don't apply to both.

File: include/interface/commandStructure.h
Comment (r-owen) [inline include/interface/commandStructure.h]: Are you actually using this value? I couldn't find it in the code.

**Review by r-owen — APPROVED** [review]
Thanks for the updates. I'm very glad you were willing to remove the sync pattern.

A few very minor comments I hope you will consider before merging.

File: include/interface/commandStructure.h
Comment (teweitsai) [reply include/interface/commandStructure.h]: The `ts_hexapod_controller` code will issue the command by itself to hack the Simulink model. For example:
https://github.com/lsst-ts/ts_hexapod_controller/blob/develop/src/commanding.c#L299-L309

Therefore, I do need this enum item before I fixed the Simulink model of hexapod. Thanks!

File: doc/commandStatus.md
Comment (teweitsai) [reply doc/commandStatus.md]: Done in the commit: cd5eb1d. Thanks!

File: doc/commandStatus.md
Comment (teweitsai) [reply doc/commandStatus.md]: Removed in commit: cd5eb1d. Thanks!

**Comment by teweitsai** [comment]
The enum values begin from 1 now in commit: cd5eb1d. Thanks!

**Pull request merged**
